### PR TITLE
Remove cloud friend from storage

### PR DESCRIPTION
**Pull request by gitlaura**
- Updated removeUser method to remove a cloud friend from storage

---

This change is [Reviewable](https://reviewable.io/reviews/uproxy/uproxy-lib/366)

### CODE REVIEW

File: src/cloud/social/provider.ts [hunk 1]
```diff
@@ -391,9 +391,19 @@ export class CloudSocialProvider {
         new Error('blockUser unimplemented'));
   }
 
-  public removeUser = (userId: string): Promise<void> => {
-    return Promise.reject(
-        new Error('removeUser unimplemented'));
+  // Removes a cloud contact from storage
+  public removeUser = (host: string): Promise<void> => {
+    log.debug('removing cloud contact %1 from storage %2', host, STORAGE_KEY);
```
Comment (trevj) [inline src/cloud/social/provider.ts]: The other social2 methods just report the name of the method being called, e.g.
```typescript
  log.debug('removeUser %1', host);
```

File: src/cloud/social/provider.ts [hunk 2]
```diff
@@ -391,9 +391,19 @@ export class CloudSocialProvider {
...
+    if (!this.savedContacts_[host]) {
```
Comment (trevj) [inline src/cloud/social/provider.ts]: I forget the exact reasons why it's safer to do:
```typescript
  if (!(host in this.savedContacts_)) {
    ...
  }
```

File: src/cloud/social/provider.ts [hunk 3]
```diff
@@ -391,9 +391,19 @@ export class CloudSocialProvider {
...
+      // Do not return an error because result is as expected.
+      log.error('cloud contact %1 is not in %2 - cannot remove from storage', host, STORAGE_KEY);
```
Comment (trevj) [inline src/cloud/social/provider.ts]: I would make this a warn, since it seems very non-fatal.

File: src/cloud/social/provider.ts [hunk 4]
```diff
@@ -391,9 +391,19 @@ export class CloudSocialProvider {
...
+    } else {
+      // Remove host from this.savedContacts_
+      delete this.savedContacts_[host];
```
Comment (trevj) [inline src/cloud/social/provider.ts]: Let's delete `this.clients_[host]`, too (no warning if it doesn't exist).

File: src/cloud/social/provider.ts [hunk 5]
```diff
@@ -391,9 +391,19 @@ export class CloudSocialProvider {
...
+      // Update storage with this.savedContacts_
+      this.saveContacts_();
+    }
+    return Promise.resolve<void>();
```
Comment (trevj) [inline src/cloud/social/provider.ts]: I was going to suggest you just `return this.saveContacts_()` here and realised I implemented a classic "swallow and log" anti-pattern...my bad. Do you mind changing `saveContacts_` to read:
```typescript
  // Saves contacts to storage.
  private saveContacts_ = () => {
    log.debug('saveContacts');
    return this.storage_.set(STORAGE_KEY, JSON.stringify(<SavedContacts>{
      contacts: Object.keys(this.savedContacts_).map(key => this.savedContacts_[key])
    });
  }
```
And **then** have this new function `return this.saveContacts_`?

Comment (gitlaura) [inline src/cloud/social/provider.ts]: Done!

File: src/cloud/social/provider.ts (hunk 4 above)
Comment (gitlaura) [inline src/cloud/social/provider.ts]: Done

File: src/cloud/social/provider.ts (hunk 3 above)
Comment (gitlaura) [inline src/cloud/social/provider.ts]: Done

File: src/cloud/social/provider.ts (hunk 2 above)
Comment (gitlaura) [inline src/cloud/social/provider.ts]: Got it. Done.

File: src/cloud/social/provider.ts (hunk 1 above)
Comment (gitlaura) [inline src/cloud/social/provider.ts]: Okay, done.

**Comment by trevj** [comment]
:+1:

**Pull request merged**